ESQL: optimise aggregations filtered by false/null into evals

bpintea opened the pull request:
This adds a new optimiser rule to extract aggregate functions filtered by a `FALSE` or `NULL` into evals. The value taken by the evaluation is `0L`, for `COUNT()` and `COUNT_DISTINCT()`, `NULL` otherwise.

Example:
```
... | STATS x = someAgg(y) WHERE FALSE {BY z} | ...
=>
... | STATS x = someAgg(y) {BY z} > | EVAL x = NULL | KEEP x{, z} | ...
```

Related: #114352.

bpintea committed:
Optimize false filters into evals

bpintea committed:
treat count_distinct as count

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

elasticsearchmachine commented:
Hi @bpintea, I've created a changelog YAML for you.

Commit:
Update docs/changelog/115858.yaml

bpintea committed:
Merge branch 'main' into enh/agg_filter_as_eval

bpintea committed:
Merge branch 'main' into enh/agg_filter_as_eval

bpintea committed:
Merge branch 'main' into enh/agg_filter_as_eval

bpintea committed:
Merge branch 'main' into enh/agg_filter_as_eval

bpintea committed:
Merge branch 'enh/agg_filter_as_eval' of github.com:bpintea/elasticsearch into enh/agg_filter_as_eval

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into enh/agg_filter_as_eval

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceStatsFilteredAggWithEval.java:
```diff
@@ -0,0 +1,65 @@
+                Object value = aggFunction instanceof Count || aggFunction instanceof CountDistinct ? 0L : null;
+                Alias newAlias = alias.replaceChild(Literal.of(aggFunction, value));
+                newEvals.add(newAlias);
+                newProjections.add(newAlias.toAttribute());
```
I got different results between the branch and main for a query like below, I wonder if we should keep the aggregation here? 🤔

main
```
+ Q2_FalseFilter
{
  "query": "from sample_data | stats c = count(*) where false, s = sum(event.duration) where false"
}
'
       c       |       s
---------------+---------------
0              |null
+ Q3_NullFilter
{
  "query": "from sample_data | stats c = count(*) where null, s = sum(event.duration) where null"
}
'
       c       |       s
---------------+---------------
0              |null
+ Q4_TrueFilter
{
  "query": "from sample_data | stats c = count(*) where true, s = sum(event.duration) where true"
}
'
       c       |       s
---------------+---------------
7              |23231327
```


branch
```
+ Q2_FalseFilter
{
  "query": "from sample_data | stats c = count(*) where false, s = sum(event.duration) where false"
}
'
       c       |       s
---------------+---------------
0              |null
0              |null
0              |null
0              |null
0              |null
0              |null
0              |null
+ Q3_NullFilter
{
  "query": "from sample_data | stats c = count(*) where null, s = sum(event.duration) where null"
}
'
       c       |       s       
---------------+---------------
0              |null           
0              |null        
0              |null          
0              |null
0              |null
0              |null           
0              |null           
+ Q4_TrueFilter
{ 
  "query": "from sample_data | stats c = count(*) where true, s = sum(event.duration) where true"
}              
'
       c       |       s
---------------+---------------
7              |23231327       
```

costin replied:
I think the aggregate gets replaced with a projection that returns the constant for each entry:
`FROM sample__data | stats c = count(*) where null, sum = sum() where null  | eval c = 0, sum = null`
becomes
`FROM sample_data | eval c = 0, sum = null`

bpintea replied:
Thanks, indeed, the case where the Aggregation was entirely pruned wasn't handled correctly. It's been fixed  now.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceStatsFilteredAggWithEval.java:
```diff
@@ -0,0 +1,65 @@
+            plan = newEvals.size() < aggregate.aggregates().size() - aggregate.groupings().size()
+                ? aggregate.with(aggregate.child(), aggregate.groupings(), newAggs)
+                : aggregate.child();
+            plan = new Eval(aggregate.source(), plan, newEvals);
+            plan = new Project(aggregate.source(), plan, newProjections);
```
I believe the eval can be skipped and instead just use the project with aliases to the underlying literal.
Can you confirm that CombineProjection eliminates the underlying aggregation (count, max, etc...)?
e.g.:
```
FROM x | STATS c = COUNT() WHERE FALSE BY y  
// becomes
FROM x | STATS c = COUNT() WHERE FALSE by y | EVAL c = 0
// which is optimized to
FROM x |  STATS BY y | EVAL c= 0
```

bpintea replied:
> I believe the eval can be skipped and instead just use the project with aliases to the underlying literal.

Not sure if this is possible without updating the plan verification: `PlanConsistencyChecker#checkPlan()` would no longer find the references, since `Alias` would delegate to `Literal`, which returns empty refs. Also, would this not prevent some other rules to apply, like `PropagateEvalFoldables`?

> Can you confirm that CombineProjection eliminates the underlying aggregation (count, max, etc...)?

Yes, the underling aggregation is being replaced ([test here](https://github.com/elastic/elasticsearch/pull/115858/files#diff-427544eaca2f06adb1b5850278a22573fb696564260a34a8f5580ff66bbbd95aR821)).

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
Please add some tests for stats with a grouping key.

bpintea replied:
There has been [one already](https://github.com/elastic/elasticsearch/pull/115858/files#diff-427544eaca2f06adb1b5850278a22573fb696564260a34a8f5580ff66bbbd95aR759), added [more](https://github.com/elastic/elasticsearch/pull/115858/files#diff-427544eaca2f06adb1b5850278a22573fb696564260a34a8f5580ff66bbbd95aR821).

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceStatsFilteredAggWithEval.java:
```diff
@@ -0,0 +1,65 @@
+public class ReplaceStatsFilteredAggWithEval extends OptimizerRules.OptimizerRule<Aggregate> {
+    @Override
+    protected LogicalPlan rule(Aggregate aggregate) {
+        List<NamedExpression> newAggs = new ArrayList<>(aggregate.aggregates().size());
```
I would make it a bit more readable and extract `aggregate.aggregates().size()` into a variable to be reused.

bpintea replied:
Makes sense, updated.

astefan reviewed (commented):
I think the rule is ok on the surface, but the lack of csv-spec tests is a bit of a concern.
For example,
- `row x = null, a = 1, b = [2,3,4] | stats c=max(a) where x by b` doesn't return a correct result. `b` should be "expanded" and there should be three rows, with this PR the returned row is `null           |[2, 3, 4]`
- `count(*)` and `count_distinct(*)` should also be covered
- maybe try other aggregations as well, `min` and `max` are the more known and seasoned ones but we do have some exotics in the list of aggs as well
- try some queries where the condition is `eval`ed in a previous command and to be used the agg filter. `eval x = null + 1 | stats ... where x + 2`. It is likely you encounter errors in some use cases, but at least make sure the PR doesn't change a working use case into a non-working one.

bpintea committed:
Add STATS pruning

bpintea committed:
Merge branch 'main' into enh/agg_filter_as_eval

bpintea committed:
spotless

bpintea committed:
add more tests

bpintea committed:
more sorting in tests

fang-xing-esql commented:
I came across this recently, it is not within the scope of this PR I think, as it is not what it tries to address. But the null filter behave differently when it is not under an aggregation. If we want to open up null filter not under aggregations, it can be addressed separately.

NullFilter is not allowed when it is not under aggregation
```
{"query": "from sample_data | where null"}
{
  "error" : {
    "type" : "verification_exception",
    "reason" : "Found 1 problem\nline 1:26: Condition expression needs to be boolean, found [NULL]"
  },
  "status" : 400
}
```
NullFilter is allowed under aggregation
```
{"query": "from sample_data | stats s = sum(event.duration) where null"}
       s       
---------------
null  
```

bpintea committed:
Merge branch 'main' into enh/agg_filter_as_eval

bpintea committed:
Skip applying rule locally

bpintea committed:
Merge branch 'main' into enh/agg_filter_as_eval

bpintea commented:
> the lack of csv-spec tests is a bit of a concern.

I've added CSV tests now too. CSVs aren't very relevant with pure optimisations, as the results wouldn't be changed, but added some more.

> * row x = null, a = 1, b = [2,3,4] | stats c=max(a) where x by b doesn't return a correct result. b should be "expanded" and there should be three rows, with this PR the returned row is null           |[2, 3, 4]

Fixed now (was same reason as above, incorrect Aggregate node pruning), added this as test too.

> * count(*) and count_distinct(*) should also be covered

These have been covered in the unit tests, now added them to the CSVs too.

> * maybe try other aggregations as well, min and max are the more known and seasoned ones but we do have some exotics in the list of aggs as well

Added.

> * try some queries where the condition is evaled in a previous command and to be used the agg filter. eval x = null + 1 | stats ... where x + 2. It is likely you encounter errors in some use cases, but at least make sure the PR doesn't change a working use case into a non-working one.

This won't make a difference currently, since PropagateEvalFoldables is applied to Aggs yet. But added some tests for when they'll be.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalLogicalPlanOptimizer.java:
```diff
@@ -50,20 +52,31 @@ protected List<Batch<LogicalPlan>> batches() {
+                if (r instanceof PropagateEmptyRelation) {
+                    newRules.add(new LocalPropagateEmptyRelation());
+                    updated = true;
+                } else if (r instanceof ReplaceStatsFilteredAggWithEval) {
```
Before disabling this rule locally, some BWC tests failed. The reason being that an old node was sending a fragment to a new one; the new node optimised locally pruning away the `Aggregate` node in the logical plan, which resulted in failed queries.
Having the right test query triggered this, but made me wonder if we could make this more resilient, since no other check caught this issue.
Also, likely other rules executed locally from `LocalPlanOptimizer.operators()` could be skipped locally (pruning, logical manipulations, maybe _some_ of the pushing etc.)

astefan replied:
Agree, but what can we do better other than BWC checks?
An alternative would be to selectively (explicitly) add rules to the local optimization batch. The upside is **awareness** (a bwc test fails -> check the newly added rules or changing an existent one must be added to LocalOptimizer or not),  the downside is explicitly reasoning about the need on the Local node as well.

Which query made this expose the issue?
I imagine it has to be a pipeline breaker. Maybe we need to get better at testing queries with pipeline breakers.

bpintea replied:
> Which query made this expose the issue?

All the newly added queries that use an index (non-ROW). These were creating a plan on the old nodes that still required sending a fragment to data nodes (running the "new" optimising version).

> I imagine it has to be a pipeline breaker.

Yes, but specifically for Aggs, with the intermediate attributes, that's particularly "sensitive". I think reoptimising the other  pipeline breakers wouldn't alter the output.

> what can we do better other than BWC checks?

It'd probably suffice to check that the output stays unchanged.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalLogicalPlanOptimizer.java:
```diff
@@ -50,20 +52,31 @@ protected List<Batch<LogicalPlan>> batches() {
     }
 
+    @SuppressWarnings("unchecked")
     private List<Batch<LogicalPlan>> replaceRules(List<Batch<LogicalPlan>> listOfRules) {
```
`listOfBatches` I'd say.

bpintea replied:
This was unchanged, but yes, good point. I'll try to pick this up with a next update, I'd like to try adding [a verification](https://github.com/elastic/elasticsearch/pull/115858/files#r1836931039) after the local plan optimisation too.

astefan reviewed (approved):
LGTM
